### PR TITLE
Upgrade frontend libraries

### PR DESCRIPTION
**Pull request by kxmbrian**
- Make [necessary changes](https://facebook.github.io/react/blog/2017/04/07/react-v15.5.0.html) to upgrade for React 15.5 and to prepare for React 16.
- `react-summernote` has not made the upgrade. Submitted a PR to the repo for that. Right now, it only generates warnings.
- Following up to #2278, removed `timeZone` prop is no longer required as a prop to `IntlProvider`. Switched to using `yahoo/react-intl` instead of `jeremyyap/react-intl`.

### CODE REVIEW

**Comment by allenwq** [comment]
👍  @kxmbrian Does the libraries in the webpack config need to be updated ?

**Comment by kxmbrian** [comment]
Update: `react-summernote` PR has been merged, but maintainer has not published the latest package.  

Latest changes:
- Updated the webpack config from [webpack 1 to webpack 2 syntax](https://webpack.js.org/guides/migrating/). 
- [Use `import` + `webpackChunkName` instead of `require.ensure`](https://webpack.js.org/guides/code-splitting-async/#require-ensure-).
- Removed some externals from webpack config - even though it is the [recommended setup for enzyme](https://github.com/airbnb/enzyme/blob/master/docs/guides/webpack.md), we don't seem to need it for your `jest` setup, i.e. removing it does not results a [`Error: Cannot resolve module...`](https://github.com/airbnb/enzyme/issues/47#issuecomment-207498885).
- Upgraded yarn version in package.json.
- `css-loader`, `jest `, `redux-logger	`, `glob`, `react-redux `, `style-loader` versions have moved since PR was first made.

Note: you might have to reinstall the node-sass package on your local machine if you get [this](https://github.com/sass/node-sass/issues/1579#issuecomment-227662011) issue. I did `yarn remove` and `yarn add`, but you can try [`yarn install --force`](https://yarnpkg.com/en/docs/cli/install#toc-yarn-install-force)

File: client/app/index.js [hunk 1]
```diff
@@ -35,12 +35,9 @@ function loadModules() {
...
-    require('intl');
-    require('intl/locale-data/jsonp/en');
-    require('intl/locale-data/jsonp/zh');
-    loadModules();
-  }, 'intl');
-} else {
-  loadModules();
+  import(/* webpackChunkName: "intl" */ 'intl');
+  import(/* webpackChunkName: "intl" */ 'intl/locale-data/jsonp/en');
+  import(/* webpackChunkName: "intl" */ 'intl/locale-data/jsonp/zh');
```
Comment (allenwq) [inline client/app/index.js]: @kxmbrian Is the import here async ?  If it is then we will need to `load_modules` after import success.

Comment (kxmbrian) [inline client/app/index.js]: Updated!

**Comment by kxmbrian** [comment]
@weiqingtoh Have made another PR to `react-summernote` to expedite the update, however, even if it were merged, we will still get warnings because the latest build of `react-summernote` uses Promises but [phantomjs does not support Promises natively](https://github.com/ariya/phantomjs/issues/12401).

It is possible to polyfill it for the time being, but it might be better to swap out phantomjs instead #2270. Shall we leave it for now and use the production build instead of the test build when testing locally?

Related to https://github.com/Coursemology/coursemology2/issues/2307#issuecomment-304599749.

**Comment by weiqingtoh** [comment]
@kxmbrian My main concern is whether or not we are able to catch critical JS errors using the production build -  because it does ignore some js errors after all. Would it be possible to study that? 

The move to chrome headless might have to wait a little cause it still requires for us to manually build chrome locally, and I'm not too keen to impose that on the development team yet. If the polyfill does not take a significant amount of time, I think it's all right to do that first.

**Comment by kxmbrian** [comment]
Yup, I agree that we should let all JS warnings come to light and we should wait a while before moving to headless chrome.

Made a PR for the polyfill #2309

**Comment by weiqingtoh** [comment]
:+1: Thanks @kxmbrian !

**Pull request merged**